### ISSUE TITLE
Infinite recursion detection during resource and scene saving.

### ISSUE THREAD

**Issue by lyuma** [godotengine/godot-proposals#2657]
### Describe the project you are working on
Custom godot resources

### Describe the problem or limitation you are having in your project
Due to engine bugs, or due to user error, it is quite common for me to trigger a mistake where a resource, Array or Dictionary references itself, directly or indirectly (e.g. via a Dictionary).

When this happens, the engine functions normally until I save the scene or modified resources, at which point it triggers infinite recursion and crashes.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
The feature is early detection for the cases that currently crash the editor (or runtime): keeping a sort of visited set.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
This can be done with a simple stack + hashtable detection for infinite recursion when saving resources. Simply put, if I recurse into a resource in _find_resources, I push it onto the stack (+hashtable).

If the engine hits a resource that is already in the stack, we abort noisily with an error but avoid corrupting the resource or crashing the engine.

### If this enhancement will not be used often, can it be worked around with a few lines of script?
A checker script could be constructed, but there's no mechanism in GDScript to easily compare dictionary and array references,

### Is there a reason why this should be core and not an add-on in the asset library?
ResourceSaver is part of core, so this should be too.

Godot engine should not be possible to crash by something as simple as dragging a resource reference into another resource.

**Comment by Xrayez** [comment]
Yeah, this kind of protection mechanism could make features like godotengine/godot#33824, godotengine/godot#33137 more safe to implement for recursively duplicating resources, for instance.
